### PR TITLE
podlogstream: fix an update loop

### PR DESCRIPTION
**Pull request by nicks**
Hello @milas,

Please review the following commits I made in branch nicks/update:

904588f5393c41a44b419dc1d86c260b80eae5b1 (2021-04-27 15:52:09 -0400)
podlogstream: fix an update loop

Code review reminders, by giving a LGTM you attest that:

* Commits are adequately tested
* Code is easy to understand and conforms to style guides
* Incomplete code is marked with TODOs
* Code is suitably instrumented with logging and metrics

### CODE REVIEW

**Comment by nicks** [comment]
gahhhhh, i wonder if we should add some sort of integration test or metric for the number of updates, catching this class of error is a pain...

File: internal/controllers/core/podlogstream/podlogstreamcontroller.go [hunk 1]
```diff
@@ -433,9 +435,16 @@ func (r *Controller) updateStatus(streamName types.NamespacedName) {
 		return
 	}
 
+	lastUpdate, hasLastUpdate := r.lastUpdate[streamName]
+	if hasLastUpdate && cmp.Equal(status, lastUpdate) {
+		return
+	}
+
+	r.lastUpdate[streamName] = status.DeepCopy()
```
Comment (milas) [inline internal/controllers/core/podlogstream/podlogstreamcontroller.go:443]: Shouldn't this only be updated if the status update actually succeeds below?

File: internal/controllers/core/podlogstream/podlogstreamcontroller.go [hunk 2]
```diff
@@ -433,9 +435,16 @@ func (r *Controller) updateStatus(streamName types.NamespacedName) {
 		return
 	}
 
+	lastUpdate, hasLastUpdate := r.lastUpdate[streamName]
+	if hasLastUpdate && cmp.Equal(status, lastUpdate) {
```
Comment (milas) [inline internal/controllers/core/podlogstream/podlogstreamcontroller.go:439]: I realize this is actually the same as before, but any reason to use `cmp` vs the K8s `equality.Semantic.DeepEqual()`?

**Review by milas — COMMENTED** [review]
I think the last update tracking needs to move to only be on success or we might miss some updates?

Also, would it make sense to, instead of storing copies of the last update result, compare to the entity you fetch?

```go
stream := &PodLogStream{}
err := r.client.Get(r.ctx, streamName, stream)
if err != nil {
	return
}
if equality.Semantic.DeepEqual(status, stream.Status) {
	return
}
// do update...
```

**Comment by nicks** [comment]
Re: would it make sense to, instead of storing copies of the last update result, compare to the entity you fetch?

Nope! This is precisely the bug. The entity you fetch might be arbitrarily different than the entity you sent.

File: internal/controllers/core/podlogstream/podlogstreamcontroller.go (hunk 1 above)
Comment (nicks) [reply internal/controllers/core/podlogstream/podlogstreamcontroller.go:443]: done

File: internal/controllers/core/podlogstream/podlogstreamcontroller.go (hunk 2 above)
Comment (nicks) [reply internal/controllers/core/podlogstream/podlogstreamcontroller.go:439]: done

**Comment by nicks** [comment]
fwiw, i do think we need to investigate other kubebuilder / controller patterns for making these updates more "natural". They're tricky to get right. "Don't read status when updating status" is a general good rule of thumb

**Review by milas — APPROVED** [review]

**Pull request merged**